Fixes to i2c and gpio drivers

mferland opened the pull request:
I found two major bugs in the i2c and gpio drivers:

- the i2c driver was corrupting the stack, which crashed the kernel
- the gpio driver was not setting the pin level when setting both the direction _and_ the level

The other fixes are mainly small improvements.

mferland committed:
i2c: make eiois200_dev a local variable

This reduces the scope and makes the variable local. There's no need
to make this variable global since it is used in a single
function (load_i2c).

mferland committed:
i2c: return ENODEV if regmap is not found

ENOMEM should only be returned when memory allocation fails, follow
what other drivers are doing and return -ENODEV.

mferland committed:
i2c: make the driver compatible with i2c_add_numbered_adapter

Make this driver allocate numbered i2c buses from 0 to 3 (incl.) to
make writing static board info structures possible. Without this
patch, the i2c adapters are dynamically numbered and it is impossible
to call the i2c_register_board_info function from board init files.

See drivers/i2c/i2c-boardinfo.c for more info.

mferland committed:
i2c: fix addr debug message

Without this patch, the address is always shown as '0'.

mferland committed:
i2c: fix stack corruption

The I2C_READ macro uses regmap_read for register access. This function
has the following signature:

    int regmap_read(struct regmap *map, unsigned int reg, unsigned int *val);

Note that the 'val' pointer is of type 'unsigned int' _but_ the
read_data function passes a 'void *', discarding the type of the
original buffer (u8 *). This will, in certain cases, make the kernel
panic because of stack corruption. In our case, we get the
following message:

    Kernel panic - not syncing: stack-protector: Kernel stack is corrupted in: mxt_t6_command+0xd4/0xe0 [atmel_mxt_ts]

Call overview:

    #define I2C_READ(i2c, offset, val)	(regmap_read(regmap, \
					 (i2c)->base + (offset), val)) <-- 3

    static int read_data(struct dev_i2c *i2c, void *data)
    {
        ...
	I2C_READ(i2c, reg, data);  <-- 2

        return 0;
    }

    static int i2c_xfer(struct i2c_adapter *adap, struct i2c_msg *msgs, int num)
    {
        ...
        for (msg = 0; msg < num; msg++)	{
            ...
            /* Transmit all messages */
            for (data = 0; data < msgs[msg].len; data++) {
                if (msgs[msg].flags & I2C_M_RD) {
                    bool last = msgs[msg].len == data + 1;

                    if (last)
                        let_stop(i2c);

                    ret = read_data(i2c, &msgs[msg].buf[data]);  <-- 1
                    ....
                }
            }
        }
    }

1- The 'buf' member of the 'struct i2c_msg' is of type 'u8*';

2- It then gets converted to 'void*' in read_data;

3- It finally gets interpreted as 'unsigned int *' in regmap_read;

Ultimately, regmap_read writes overflow in the stack of the calling
function and make the kernel crash.

To fix this:

- stop converting the buffer to void;
- use a temporary variable for the regmap_read call;

mferland committed:
gpio: return -ENODEV if device cannot be probed

mferland committed:
gpio: fix gpio access functions

The dir_output function does not take the value of the
output (high/low) into account. This leads to situations where the
initial state is wrong.

Also, the gpio_set function passes a pointer to an 'int' and then uses
it as if it were a pointer to a 'u8' inside of pmc_write. This should
be avoided.

wenkaichung reviewed (approved):
> ENOMEM should only be returned when memory allocation fails, follow
> what other drivers are doing and return -ENODEV.
It is inappropriate to return ENOMEM

mferland commented:
Is there anything I can do so these commits can be merged? Thanks!

wenkaichung commented:
Hi Marc,

Thank you for your contribution and patience. We appreciate your valuable patch.
We are currently scheduling the patch verification. Once the QA testing is
completed successfully, we'll proceed with merging your patch into the mainline.

Thank you again for your valuable contribution.

Sincerely,

Wenkai Chung

Tel: +886-2-2792-7818 ext#: 9073
Advantech Co., Ltd.
[2011_Emb'Core]

From: Marc Ferland ***@***.***>
Sent: Tuesday, April 16, 2024 12:01 AM
To: ADVANTECH-Corp/eio-is200-linux-kernel-driver ***@***.***>
Cc: Wenkai.Chung ***@***.***>; Comment ***@***.***>
Subject: Re: [ADVANTECH-Corp/eio-is200-linux-kernel-driver] Fixes to i2c and gpio drivers (PR #2)


Is there anything I can do so these commits can be merged? Thanks!

—
Reply to this email directly, view it on GitHub<https://apc01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FADVANTECH-Corp%2Feio-is200-linux-kernel-driver%2Fpull%2F2%23issuecomment-2057204059&data=05%7C02%7CWenkai.Chung%40advantech.com.tw%7C438fecba7e6b457dc80808dc5d6534a0%7Ca77d40d9dcba4ddab5715f18e6da853f%7C0%7C0%7C638487936476284046%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C0%7C%7C%7C&sdata=FLOAZdPbQl%2BwhvhwGyvYHmjhT7xkRkoR24ie5VT8mW4%3D&reserved=0>, or unsubscribe<https://apc01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fnotifications%2Funsubscribe-auth%2FBHWCTZZAYYSCVNL3JVNLKEDY5P2SRAVCNFSM6AAAAABFUABFUKVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDANJXGIYDIMBVHE&data=05%7C02%7CWenkai.Chung%40advantech.com.tw%7C438fecba7e6b457dc80808dc5d6534a0%7Ca77d40d9dcba4ddab5715f18e6da853f%7C0%7C0%7C638487936476284046%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C0%7C%7C%7C&sdata=8mTxqsrOoAcThj7RqN8e06vnQ1JFUztgskFpjT0yyZY%3D&reserved=0>.
You are receiving this because you commented.Message ID: ***@***.******@***.***>>

mferland committed:
thermal: fix temperature conversion

The kernel typically uses milli-celcius to represent temperatures. So
for example 25000 is 25C. Currently the thermal driver converts to
deci-celcius. Fix this by rewriting the conversion macros.

Example from stress-ng:

stress-ng: info:  [5667] setting to a 300 second (5 mins, 0.00 secs) run per stressor
stress-ng: info:  [5667] dispatching hogs: 16 matrix
stress-ng: info:  [5667] matrix:
stress-ng: info:  [5667]               acpitz   41.30 C (314.45 K)
stress-ng: info:  [5667]         eiois200_fan0   41.30 C (314.45 K)
stress-ng: info:  [5667]         eiois200_fan1   41.30 C (314.45 K)
stress-ng: info:  [5667]     eiois200_thermal    0.41 C (273.56 K)
stress-ng: info:  [5667]           sbs_5_000b   24.60 C (297.75 K)

The eiois200_thermal thermal zone reports a temperature of 0.41C at
the end of the run but it should have been 41.30C like the others.

Signed-off-by: Marc Ferland <marc.ferland@sonatest.com>

mferland committed:
thermal: set passive and polling delays to 1 and 2 sec. resp

Passing 0 to both passive and polling delays will disable the kernel
monitoring code and thus will prevent the OS from powering off the
system when the temperature reaches a critical point.

Signed-off-by: Marc Ferland <marc.ferland@sonatest.com>

mferland committed:
core: reuse regmap_read_poll_timeout instead of rolling our own logic

The regmap API provides a regmap_read_poll_timeout based of the iopoll
macros. Reuse this macro instead of reinventing the wheel.

I also discovered that the previous logic would often return errors
even though the read would have completed successfully. For example,
reading the temperature sensor would often print errors like:

[  201.429357] thermal thermal_zone3: PMC error duration:10919us
[  201.429370] thermal thermal_zone3: .cmd=0x11, .ctrl=0x10 .id=0x00, .size=0x02 .data=0x0000

The new logic seems much more stable. I wasn't able to reproduce the
read errors.

Signed-off-by: Marc Ferland <marc.ferland@sonatest.com>

mferland commented:
Hi Wenkai,

I added more patches to the pull request. Tell me if there's anything else I can do.

Marc

mferland committed:
core: fix pmc_read_function

The pmc_read_function was actually reading the _data_ port instead of
the _status_ port. Also, fix the error message.

mferland committed:
core: fix pmc_write_cmd error message

When printing the error message, we display the wrong port
address. Fix this by changing data --> cmd.

mferland committed:
eiois200_wdt: fix comment in set_time

We multiply by 1000 to convert the timeout from seconds to
milliseconds.
